File: app/src/main/java/com/apps/adrcotfas/goodtime/Util/OnSwipeTouchListener.java
/*
 * Copyright 2016-2019 Adrian Cotfas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

package com.apps.adrcotfas.goodtime.Util;

import android.content.Context;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;

public abstract class OnSwipeTouchListener implements View.OnTouchListener {

    private final GestureDetector gestureDetector;
    private View view;
    private boolean mIsScrolling;

    enum SwipeDirection {
        LEFT,
        RIGHT,
        UP,
        DOWN,
        CENTER
    }

    private SwipeDirection mDirection;

    public OnSwipeTouchListener(Context context) {
        gestureDetector = new GestureDetector(context, new GestureListener());
    }

    @Override
    public boolean onTouch(View view, MotionEvent event) {
        this.view = view;
        if (event.getAction() == MotionEvent.ACTION_DOWN) {
            onPress(view);
        } else if (event.getAction() == MotionEvent.ACTION_UP) {
            if(mIsScrolling) {
                mIsScrolling  = false;
                switch (mDirection) {
                    case UP:
                        onSwipeTop(view);
                        break;
                    case DOWN:
                        onSwipeBottom(view);
                        break;
                    case LEFT:
                        onSwipeLeft(view);
                        break;
                    case RIGHT:
                        onSwipeRight(view);
                        break;
                    default:
                        break;
                }
            }
            onRelease(view);
        }
        return gestureDetector.onTouchEvent(event);
    }

    protected abstract void onSwipeRight(View view);
    protected abstract void onSwipeLeft(View view);
    protected abstract void onSwipeBottom(View view);
    protected abstract void onSwipeTop(View view);
    protected abstract void onClick(View view);
    protected abstract void onLongClick(View view);
    protected abstract void onPress(View view);
    protected abstract void onRelease(View view);

    private final class GestureListener extends GestureDetector.SimpleOnGestureListener {

        private static final int SWIPE_THRESHOLD = 150;

        @Override
        public boolean onDown(MotionEvent e) {
            return true;
        }

        @Override
        public void onLongPress(MotionEvent e) {
            onLongClick(view);
            super.onLongPress(e);
        }

        @Override
        public boolean onSingleTapUp(MotionEvent e) {
            onClick(view);
            return super.onSingleTapUp(e);
        }

        @Override
        public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float distanceY) {
            boolean result = false;
            try {
                float diffY = e2.getY() - e1.getY();
                float diffX = e2.getX() - e1.getX();
                if (Math.abs(diffX) > Math.abs(diffY)) {
                    if (Math.abs(diffX) > SWIPE_THRESHOLD) {
                        mIsScrolling = true;
                        if (diffX > 0) {
                            mDirection = SwipeDirection.RIGHT;
                        } else {
                            mDirection = SwipeDirection.LEFT;
                        }
                        result = true;
                    }
                }
                else if (Math.abs(diffY) > SWIPE_THRESHOLD) {
                    mIsScrolling = true;
                    if (diffY > 0) {
                        mDirection = SwipeDirection.DOWN;
                    } else {
                        mDirection = SwipeDirection.UP;
                    }
                    result = true;
                } else {
                    mDirection = SwipeDirection.CENTER;
                    mIsScrolling = true;
                }
            } catch (Exception exception) {
                exception.printStackTrace();
            }
            return result;
        }
    }
}

File: app/src/main/java/com/apps/adrcotfas/goodtime/util/OnSwipeTouchListener.kt
/*
 * Copyright 2016-2019 Adrian Cotfas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */
package com.apps.adrcotfas.goodtime.util

import android.content.Context
import android.view.View.OnTouchListener
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.View
import java.lang.Exception
import kotlin.math.abs

abstract class OnSwipeTouchListener(context: Context?) : OnTouchListener {

    private val gestureDetector: GestureDetector
    private lateinit var view: View

    override fun onTouch(view: View, event: MotionEvent): Boolean {
        this.view = view
        if (event.action == MotionEvent.ACTION_DOWN) {
            onPress(view)
        } else if (event.action == MotionEvent.ACTION_UP) {
            onRelease(view)
        }
        return gestureDetector.onTouchEvent(event)
    }

    protected abstract fun onSwipeRight(view: View)
    protected abstract fun onSwipeLeft(view: View)
    protected abstract fun onSwipeBottom(view: View)
    protected abstract fun onSwipeTop(view: View)
    protected abstract fun onClick(view: View)
    protected abstract fun onLongClick(view: View)
    protected abstract fun onPress(view: View)
    protected abstract fun onRelease(view: View)
    private inner class GestureListener : SimpleOnGestureListener() {
        override fun onDown(e: MotionEvent): Boolean {
            return true
        }

        override fun onLongPress(e: MotionEvent) {
            onLongClick(view)
            super.onLongPress(e)
        }

        override fun onSingleTapUp(e: MotionEvent): Boolean {
            onClick(view)
            return super.onSingleTapUp(e)
        }

        override fun onFling(
            e1: MotionEvent,
            e2: MotionEvent,
            velocityX: Float,
            velocityY: Float
        ): Boolean {
            var result = false
            try {
                val diffY = e2.y - e1.y
                val diffX = e2.x - e1.x
                if (abs(diffX) > abs(diffY)) {
                    if (abs(diffX) > SWIPE_THRESHOLD && abs(velocityX) > SWIPE_VELOCITY_THRESHOLD) {
                        if (diffX > 0) {
                            onSwipeRight(view)
                        } else {
                            onSwipeLeft(view)
                        }
                        result = true
                    }
                } else if (abs(diffY) > SWIPE_THRESHOLD && abs(velocityY) > SWIPE_VELOCITY_THRESHOLD) {
                    if (diffY > 0) {
                        onSwipeBottom(view)
                    } else {
                        onSwipeTop(view)
                    }
                    result = true
                }
            } catch (exception: Exception) {
                exception.printStackTrace()
            }
            return result
        }
    }

    companion object {
        private const val SWIPE_THRESHOLD = 100
        private const val SWIPE_VELOCITY_THRESHOLD = 100
    }

    init {
        gestureDetector = GestureDetector(context, GestureListener())
    }
}